R/as.data.table.R

```diff
@@ -215,7 +215,11 @@ as.data.table.list = function(x,
 
 as.data.table.data.frame = function(x, keep.rownames=FALSE, key=NULL, ...) {
   if (is.data.table(x)) return(as.data.table.data.table(x, key=key)) # S3 is weird, #6739. Also # nocov; this is tested in 2302.{2,3}, not sure why it doesn't show up in coverage.
-  if (!identical(class(x), "data.frame")) return(as.data.table(as.data.frame(x), keep.rownames=keep.rownames, key=key, ...))
+  if (!identical(class(x), "data.frame")) {
+    x = as.data.frame(x)
+    setattr(x, "class", "data.frame") # cater for cases when as.data.frame can generate a loop #6874
```

Comment:
This might not be ideal, e.g. if there's a "proper" S3 class where we have

```r
class(x) = c('class1', 'class2', 'data.frame')
class(as.data.frame(x)) = c('class2', 'data.frame')
class(as.data.frame(as.data.frame(x))) = 'data.frame'
```

Then we exit the loop, and possibly apply one additional layer of `as.data.frame()` that is required for correct results.

Unfortunately I'm not sure right workaround? Maybe record `class_orig = class(x); x=as.data.frame(x); if (identical(class(x), class_orig)) setattr(x, 'class', 'data.frame')`?

Maybe moot in practice, let's monitor the revdeps.

Reply:
I added the check. Better to be safe than sorry and as we saw some objects apparently like this nested `as.data.frame` conversion.